cmd/juju/application/deployer/bundlehandler.go

```diff
@@ -380,6 +377,76 @@ func (h *bundleHandler) resolveCharmsAndEndpoints() error {
 	return nil
 }
 
+func (h *bundleHandler) resolveCharmChannelAndRevision(charmURL, charmSeries, charmChannel, arch string) (string, int, error) {
+	if h.isLocalCharm(charmURL) {
+		return charmChannel, -1, nil
+	}
+	// If the charm URL already contains a revision, return that before
+	// attempting to resolve a revision from any charm store. We can ignore the
+	// error here, as we want to just parse out the charm URL.
+	// Resolution and validation of the charm URL happens further down.
+	curl, err := charm.ParseURL(charmURL)
+	if err == nil {
```

Comment:
Minor nit; `curl` can be scoped as it is not used outside the if block.

```suggestion
	if curl, err := charm.ParseURL(charmURL); err == nil {
```